cql3/restrictions/statement_restrictions.cc

```diff
@@ -1406,8 +1398,160 @@ bool is_satisfied_by(
+        }
+    }
+
+    bytes_view_opt value() const {
+        if (_unbounded) {
+            throw std::logic_error("bound_t::value() called on unbounded");
+        }
+        return _value;
+    }
+
+private:
+    /// If the comparison *this<=>that can be shortcircuited (due to unbounded or null cases), returns the
+    /// comparison result.  Otherwise, returns nullopt.
+    std::optional<bool> shortcircuit(const bound_t& that) const {
+        if (that._unbounded) {
+            return true; // Anything is tighter than no bound.
+        } else if (this->_unbounded || !that._value) { // null always wins: it makes the slice empty.
+            return false;
+        } else if (!this->_value) { // Now that we're sure _value is valid, check it for null.
+            return true;
+        } else {
+            return std::nullopt;
+        }
+    }
+};
+
+/// True iff op means bnd type of bound.
+bool matches(const operator_type* op, statements::bound bnd) {
+    static const std::vector<std::vector<const operator_type*>> operators{
+        {&operator_type::EQ, &operator_type::GT, &operator_type::GTE}, // These mean a lower bound.
+        {&operator_type::EQ, &operator_type::LT, &operator_type::LTE}, // These mean an upper bound.
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+bound_t get_bound(const expression& restr, const query_options& options, statements::bound bnd) {
+    return std::visit(overloaded_functor{
+            [&] (bool v) {
+                if (v) {
+                    return bound_t(*byte_type);
+                } else {
+                    throw std::logic_error("empty set has no bounds");
+                }
+            },
+            [&] (const conjunction& conj) {
+                if (conj.children.empty()) {
+                    throw std::logic_error("Empty conjunction");
+                }
+                auto invoke_get_bound = [&] (const expression& e) { return get_bound(e, options, bnd); };
+                // All conjunction elements have the same LHS; this is how
+                // single_column_restrictions::add_restriction constructs it.
+                std::vector<bound_t> children_bounds(
+                        boost::make_transform_iterator(conj.children.cbegin(), invoke_get_bound),
+                        boost::make_transform_iterator(conj.children.cend(), invoke_get_bound));
+                if (is_start(bnd)) {
+                    return *boost::max_element(children_bounds, [&] (const bound_t& a, const bound_t& b) {
+                        return b.is_tighter_lb_than(a);
+                    });
+                } else {
+                    return *boost::min_element(children_bounds, [&] (const bound_t& a, const bound_t& b) {
+                        return a.is_tighter_ub_than(b);
+                    });
+                }
```

Comment:
A tighter implementation is to use std::accumulate, with the input being the transformed thing and the oeration being min/max:

```c++
   auto op = is_start ? &bound_t::is_tighter_lb_than : &bound_t::is_tighter_ub_than;
   return boost::accumulate(conj.children
             | boost::adaptor::transformed(invoke_get_bound),
         bound_t(find the time somehow),
         std::mem_fn(op));

Reply:
(no vector is needed)

Reply:
Done.